File: 567.permutation-in-string.py
#
# @lc app=leetcode id=567 lang=python3
#
# [567] Permutation in String
#

# @lc code=start
from collections import Counter

class Solution:
    def checkInclusion(self, s1: str, s2: str) -> bool:
        
        #todo Problem decompose and understand:                                   The order has to be that s2 contains a permutation of s1 not the other way around                                                                                      how do you check that if s2 contains a permutation of s1 or not

        #! The solution is sliding window and hash map on s1 to record the unique characters and their count in s1


        # #! I missed the fact that 
        # ans = False
        # hashmap = Counter()
        # left = 0
        # #! Edge case
        # if len(s1) > len(s2):
        #     return ans 
        
        # #Generate the unique characters in s1
        # for i in range(len(s1)):
        #     hashmap[s1[i]] += 1


        # #! Sliding window       
        # for i in range(len(s2)):
            
        #     if s2[i] in hashmap:
        #         hashmap[s2[i]] -= 1
            
        #     if i < len(s1) - 1:
        #         continue

        #     count = 0
        #     for value in hashmap.values():
        #         if value != 0:
        #             count += 1
        #     if count == 0:
        #         ans = True
        #         break

        #     #! Shrink the window
        #     if s2[left] in hashmap:
        #         hashmap[s2[left]] += 1
        #     left += 1
        # return ans

        hashmap = Counter()
        for s in s1:
            hashmap[s] += 1

        left = 0
        ans = False

        for right in range(len(s2)):
            if s2[right] in hashmap:
                hashmap[s2[right]] -= 1
            
            if right < len(s1) - 1:
                continue

            if all(value == 0 for value in hashmap.values()):
                ans = True
                break
            else:
                if s2[left] in hashmap:
                    hashmap[s2[left]] += 1
                left += 1
        return ans
 # @lc code=end

File: 567.permutation-in-string.java
/*
 * @lc app=leetcode id=567 lang=java
 *
 * [567] Permutation in String
 */

// @lc code=start
class Solution {
    public boolean checkInclusion(String s1, String s2) {
        // int n1 = s1.length();
        // int n2 = s2.length();
        // boolean ans = false;
        // int[] perm = new int[26];
        // if(n1 > n2){
        //     return ans;
        // }
        // for(int i = 0; i < n1;i++){
        //     perm[s1.charAt(i) - 'a']++;
        // }
        // for(int i = 0; i < n2; i++){
        //     perm[s2.charAt(i) - 'a']--;
        //     if(i < n1 -1){
        //         continue;
        //     }
        //     if(check(perm)){
        //         ans = true;
        //         break;
        //     }
        //     perm[s2.charAt(i - n1 + 1) - 'a']++;
        // }
        // return ans; 
        int n1 = s1.length();
        int n2 = s2.length();
        int[] cnt = new int[26];
        if(n1 > n2){
            return false;
        }
        for(int i = 0; i < n1; i++){
            cnt[s1.charAt(i) - 'a']++;
        }
        for(int i = 0; i < n2; i++){
            char c = s2.charAt(i);
            cnt[c - 'a']--;
            if(i < n1-1){
                continue;
            }
            if(check(cnt)){
                return true;
            }
            cnt[s2.charAt(i - n1 + 1) - 'a']++;
        }
        return false;
    }
    private boolean check(int[] perm){
        for(int i = 0; i < perm.length; i++){
            if(perm[i] != 0){
                return false;
            }
        }
        return true;
    }
}
// @lc code=end

